crates/dprint/src/commands/lsp/mod.rs

```diff
@@ -0,0 +1,154 @@
+            // TODO: Send the actual format request.
+            let token = Arc::new(CancellationToken::new());
+            let _request = HostFormatRequest {
+              file_path,
+              file_text,
+              range,
+              override_config: Default::default(), // TODO: Set from params?
+              token,
+            };
+            // TODO: How to get the plugin scope?
+            // TODO: Return back an actual response
+            let _ = sender.send(Ok(None));
+          });
+        }
+      }
+    }
+  });
+
+  let lsp_task = dprint_core::async_runtime::spawn(async move {
+    let (service, socket) = LspService::new(|client| Backend { client, sender: tx });
+    Server::new(stdin, stdout, socket).serve(service).await;
+  });
+
+  try_join!(recv_task, lsp_task)?;
+
+  Ok(())
+}
+
+#[tower_lsp::async_trait]
+impl LanguageServer for Backend {
+  async fn initialize(&self, params: InitializeParams) -> Result<InitializeResult> {
+    if let Some(parent_id) = params.process_id {
+      start_parent_process_checker_task(parent_id);
+    }
+
+    Ok(InitializeResult {
+      // TODO: Any server info we want to include here?
+      server_info: None,
+      capabilities: ServerCapabilities {
+        text_document_sync: Some(TextDocumentSyncCapability::Options(TextDocumentSyncOptions {
+          change: Some(TextDocumentSyncKind::FULL),
+          open_close: Some(true),
+          save: None,
+          will_save: None,
+          will_save_wait_until: None,
+        })),
+        document_formatting_provider: Some(OneOf::Left(true)),
+        document_range_formatting_provider: Some(OneOf::Left(true)),
+        ..ServerCapabilities::default()
+      },
+    })
+  }
+
+  async fn initialized(&self, _: InitializedParams) {
+    self.client.log_message(MessageType::INFO, "Server initialized.").await;
+  }
+
+  async fn did_open(&self, _: DidOpenTextDocumentParams) {
+    todo!()
+  }
+
+  async fn did_change(&self, _: DidChangeTextDocumentParams) {
+    todo!()
+  }
+
+  async fn did_close(&self, _: DidCloseTextDocumentParams) {
+    todo!()
+  }
+
+  async fn formatting(&self, params: DocumentFormattingParams) -> Result<Option<Vec<TextEdit>>> {
+    let (sender, receiver) = oneshot::channel();
+    let file_path = params.text_document.uri.to_file_path().expect("Expected a valid path");
+    let range = Default::default(); // TODO: Set it based on the file length
+    let file_text = "".into(); // TODO: Fetch from the latest document sync message
+    self.sender.send(ChannelMessage::Format(file_path, range, file_text, sender)).unwrap();
+    receiver.await.unwrap()
+  }
+
+  async fn range_formatting(&self, params: DocumentRangeFormattingParams) -> Result<Option<Vec<TextEdit>>> {
+    let (sender, receiver) = oneshot::channel();
+    let file_path = params.text_document.uri.to_file_path().expect("Expected a valid path");
+    let range = Default::default(); // TODO: Map LSP range to FormatRange
```

Comment:
@dsherret what would you suggest in order to perform this mapping? Do we copy [the implementation from `deno`](https://github.com/denoland/deno/blob/b1ce2e41676ab5bc807a705b072986d9357fece5/cli/lsp/repl.rs#L259)?

Reply:
Deno internally uses this crate https://crates.io/crates/text_lines

I'm not sure it's worth computing and storing this though because it will do it for the whole document, which will probably be unnecessary. It can work for a first pass, but perhaps in the long term this could implement something custom.

Reply:
Sounds good!